src/category_theory/abelian/non_preadditive.lean

```diff
@@ -0,0 +1,727 @@
+    { intro s,
+      ext; simp, erw category.comp_id },
+    { intros s m h,
+      haveI : mono (prod.lift (𝟙 A) (0 : A ⟶ A)) := mono_of_mono_fac (prod.lift_fst _ _),
+      apply (cancel_mono (prod.lift (𝟙 A) (0 : A ⟶ A))).1,
+      convert h walking_parallel_pair.zero,
+      ext; simp } },
+  let hp2 : is_colimit (cokernel_cofork.of_π (limits.prod.snd : A ⨯ A ⟶ A) hlp),
+  { exact epi_is_cokernel_of_kernel _ hp1 },
+  apply epi_of_zero_cancel,
+  intros Z z hz,
+  have h : prod.lift (𝟙 A) (0 : A ⟶ A) ≫ cokernel.π (Δ A) ≫ z = 0,
+  { rw [←category.assoc, hz] },
+  obtain ⟨t, ht⟩ := cokernel_cofork.is_colimit.desc' hp2 _ h,
+  rw cokernel_cofork.π_of_π at ht,
+  have htt : t = 0,
+  { rw [←category.id_comp t],
+    change 𝟙 A ≫ t = 0,
+    rw [←limits.prod.lift_snd (𝟙 A) (𝟙 A), category.assoc, ht, ←category.assoc,
+      cokernel.condition, has_zero_morphisms.zero_comp] },
+  apply (cancel_epi (cokernel.π (Δ A))).1,
+  rw [←ht, htt, has_zero_morphisms.comp_zero, has_zero_morphisms.comp_zero]
+end
+
+instance is_iso_r {A : C} : is_iso (r A) :=
+is_iso_of_mono_of_epi _
+
+/-- The composite `A ⨯ A ⟶ cokernel (Δ A) ⟶ A` given by the natural projection into the cokernel
+    followed by the inverse of `r`. In the category of modules, using the normal kernels and
+    cokernels, this map is equal to the map `(a, b) ↦ a - b`, hence the name `σ` for
+    "subtraction". -/
+abbreviation σ {A : C} : A ⨯ A ⟶ A := cokernel.π (Δ A) ≫ is_iso.inv (r A)
+
+end
+
+@[simp, reassoc] lemma Δ_σ {X : C} : Δ X ≫ σ = 0 :=
+by rw [cokernel.condition_assoc, has_zero_morphisms.zero_comp]
+
+@[simp, reassoc] lemma lift_σ {X : C} : prod.lift (𝟙 X) 0 ≫ σ = 𝟙 X :=
+by rw [←category.assoc, is_iso.hom_inv_id]
+
+@[simp, reassoc] lemma Δ_map {X Y : C} (f : X ⟶ Y) : Δ X ≫ limits.prod.map f f = f ≫ Δ Y :=
+by ext; simp; erw category.id_comp
+
+@[reassoc] lemma lift_map {X Y : C} (f : X ⟶ Y) :
+  prod.lift (𝟙 X) 0 ≫ limits.prod.map f f = f ≫ prod.lift (𝟙 Y) 0 :=
+by ext; simp; erw category.id_comp
+
+/-- σ is a cokernel of Δ X. -/
+def is_colimit_σ {X : C} : is_colimit (cokernel_cofork.of_π σ Δ_σ) :=
+cokernel.cokernel_iso _ σ (as_iso (r X)).symm (by simp)
+
+/-- This is the key identity satisfied by `σ`. -/
+lemma σ_comp {X Y : C} (f : X ⟶ Y) : σ ≫ f = limits.prod.map f f ≫ σ :=
+begin
+  obtain ⟨g, hg⟩ :=
+    cokernel_cofork.is_colimit.desc' is_colimit_σ (limits.prod.map f f ≫ σ) (by simp),
+  suffices hfg : f = g,
+  { rw [←hg, cofork.π_of_π, hfg] },
+  calc f = f ≫ prod.lift (𝟙 Y) 0 ≫ σ : by rw [lift_σ, category.comp_id]
+    ... = prod.lift (𝟙 X) 0 ≫ limits.prod.map f f ≫ σ : by rw lift_map_assoc
+    ... = prod.lift (𝟙 X) 0 ≫ σ ≫ g : by rw [←hg, cokernel_cofork.π_of_π]
+    ... = g : by rw [←category.assoc, lift_σ, category.id_comp]
+end
+
+section
+
+/-- Subtraction of morphisms is defined as `(f, g)` followed by the subtraction map. -/
+def sub {X Y : C} (f g : X ⟶ Y) : X ⟶ Y := prod.lift f g ≫ σ
+
+/-- We define `-a` as `0 - a`. -/
```

Comment:
These docstrings all make sense within the context of this file, but out of context they don't really help. Again, I'm confused about what the intended scope is.

Reply:
Good points. I have attempted to improve this, what do you think?

Reply:
Yup, that looks fine to me. Thanks!